gemini/tuning/gemini_supervised_finetuning_text_classification.ipynb

```diff
@@ -0,0 +1,2295 @@
+      "execution_count": null,
+      "metadata": {
+        "id": "Fz4tGANnfcuA"
+      },
+      "outputs": [],
+      "source": [
+        "def backoff_hdlr(details):\n",
+        "    print(\"Backing off {wait:0.1f} seconds after {tries} tries \".format(**details))\n",
+        "\n",
+        "\n",
+        "def log_error(msg, *args):\n",
+        "    mp.get_logger().error(msg, *args)\n",
+        "    raise Exception(msg)\n",
+        "\n",
+        "\n",
+        "def handle_exception_threading(f):\n",
+        "    def applicator(*args, **kwargs):\n",
+        "        try:\n",
+        "            return f(*args, **kwargs)\n",
+        "        except:\n",
+        "            log_error(traceback.format_exc())\n",
+        "\n",
+        "    return applicator\n",
+        "\n",
+        "\n",
+        "@handle_exception_threading\n",
+        "@backoff.on_exception(backoff.expo, ResourceExhausted, max_tries=30, on_backoff=backoff_hdlr)\n",
+        "def _predict_message(message: str, model: GenerativeModel):\n",
+        "    \"\"\"\n",
+        "    Predict messages\n",
+        "    \"\"\"\n",
+        "    response = model.generate_content([message],\n",
+        "                        stream=False\n",
+        "                    )\n",
+        "    return response.text\n",
+        "\n",
+        "\n",
+        "def batch_predict(messages: list, model: GenerativeModel, max_workers: int=4) ->list:\n",
+        "    \"\"\"\n",
+        "    Predicts the classes for a list of messages\n",
+        "\n",
+        "    Args:\n",
+        "        - messages: list of all messages to predict\n",
+        "        - model_name: model to use for predicting.\n",
+        "        - max_workers: number of workers to use for parallel predictions\n",
+        "\n",
+        "    Returns:\n",
+        "        - list of predicted labels\n",
+        "\n",
+        "    \"\"\"\n",
+        "    predictions = list()\n",
+        "    with ThreadPoolExecutor(max_workers) as pool:\n",
+        "        partial_func = partial(\n",
+        "            _predict_message,\n",
+        "            model=model\n",
+        "\n",
+        "        )\n",
+        "        for message in tqdm(pool.map(partial_func, messages), total=len(messages)):\n",
+        "            predictions.append(message)\n",
+        "            pass\n",
+        "\n",
+        "    return predictions\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "0tKVjsJKfcuA"
+      },
+      "source": [
+        "#### Experiment Logging - Helper Functions"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "mJ5BM7Z5fcuA"
+      },
+      "outputs": [],
+      "source": [
+        "def create_experiment(\n",
+        "    experiment_name: str,\n",
+        "    project: str,\n",
+        "    location: str,\n",
+        "    experiment_description: str=\"\",\n",
```

Comment:
It's generally not recommended to pass an empty string as a default argument. Make this an `Optional[str]` type and set  a default as `None`